### PR TITLE
Feature/provider

### PR DESCRIPTION
**Pull request by Lee-Yoonjoo**
Added provider patter for WidgetList, IconList and ExpandableList

### CODE REVIEW

**Comment by mkoechig** [comment]
login.dart file is obsolete and should be deleted.
gitflow.dart consists of text only and should not be a dart file. The notes can be moved into README.md instead and also make use of the markdown formatting.

File: lib/providers/provider.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import 'package:flutter/material.dart';
+import 'package:mixeditems/models/iconitem.dart';
+import 'package:mixeditems/models/listitem.dart';
+import 'package:mixeditems/repository/repo.dart';
+
+class CustomProvider extends ChangeNotifier {
```
Comment (mkoechig) [inline lib/providers/provider.dart]: CustomProvider is not a clear name for this. Use more explicit naming, like, what is this provider providing?

File: lib/providers/provider.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  Repo _data = Repo();
```
Comment (mkoechig) [inline lib/providers/provider.dart]: It rather is a data source than actual data

File: lib/providers/provider.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    _listItems = _data.getListItems();
+  }
+
+  loadIconItems() {
+    _iconItems = _data.getIconListItems();
+  }
+
+  loadIntItems() {
+    _intItems = _data.getMockData();
+  }
```
Comment (mkoechig) [inline lib/providers/provider.dart]: Ignore this for now... placeholder^^

These should convert data to widgets. See Repo class for details.

File: lib/widgets/home.dart [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import 'package:flutter/material.dart';
+import 'package:mixeditems/providers/provider.dart';
+import 'package:provider/provider.dart';
+
+class Home extends StatelessWidget {
```
Comment (mkoechig) [inline lib/widgets/home.dart]: Dart has a coding guideline to have a keyed constructor in any widget: https://dart-lang.github.io/linter/lints/use_key_in_widget_constructors.html

Comment (mkoechig) [inline lib/widgets/home.dart]: I would suffix Widgets with a "...Widget" to make things clear

**Review by mkoechig — CHANGES_REQUESTED** [review]

**Review by mkoechig — COMMENTED** [review]
A repository should only return models, not widgets.
Widgets are view entities (presentation layer) and don't belong into a repository (domain layer).
It should be a widget's or it's provider's responsibility to take the data model and create a view/widget with the information provided within.

I will research some common provider implementations and update this later on with more information on how to use it.

**Comment by mkoechig** [comment]
This branch should not merge to master, but develop instead

File: lib/providers/provider.dart (hunk 1 above)
Comment (Lee-Yoonjoo) [reply lib/providers/provider.dart]: I think people can use different provider by data model
I will take IconListProvider for this project and handle other data on another Provider or another way. 

I have checked his provider architecture sample. https://github.com/brianegan/flutter_architecture_samples/blob/master/change_notifier_provider/lib/todo_list_model.dart

Comment (Lee-Yoonjoo) [reply lib/providers/provider.dart]: I am not so sure for now, which is the breakpoint for multiproviders .. Mit anderen Worten when we will need to create new Provider...

File: lib/providers/provider.dart (hunk 2 above)
Comment (Lee-Yoonjoo) [reply lib/providers/provider.dart]: I think that it will be better with _repo.

File: lib/widgets/home.dart (hunk 4 above)
Comment (Lee-Yoonjoo) [reply lib/widgets/home.dart]: I have changed it! :)

Comment (Lee-Yoonjoo) [reply lib/widgets/home.dart]: Oh Okay, I changed others classes names which extends StatefulWidget or StatelessWidget.

File: lib/providers/provider.dart (hunk 3 above)
Comment (Lee-Yoonjoo) [reply lib/providers/provider.dart]: I understand. I tried to keep them clear !

File: lib/providers/provider.dart (hunk 1 above)
Comment (mkoechig) [reply lib/providers/provider.dart]: As a rule of thumb: when the class get's too big :D
500 Lines of code is often considered a soft limit.

File: lib/repository/repo.dart [hunk 5]
```diff
@@ -1,34 +1,35 @@
...
+      IconItem('Emergency Information', '8 minutes ago', Icons.warning, '355.0.0.11.108'),
+      IconItem('Maps', '8 minutes ago', Icons.pin_drop,'11.18.3'),
     ];
   }
 
   List<int> getMockData() {
     return [1, 2, 3, 4, 5];
   }
+
+  List<IconDetail> getDataUsageList (){
```
Comment (mkoechig) [inline lib/repository/repo.dart]: In this class you didn't hit the auto format, did you?

File: lib/repository/widget_placeholder.dart [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import 'package:mixeditems/sections/home/icon_list_section.dart';
+import 'package:mixeditems/sections/home/info_section.dart';
+import 'package:mixeditems/sections/home/question_section.dart';
+import 'package:mixeditems/sections/home/switch_section.dart';
+import 'package:mixeditems/sections/icon_detail/head_section.dart';
+import 'package:mixeditems/sections/icon_detail/settings_section.dart';
+import 'package:mixeditems/sections/icon_detail/usage_section.dart';
+import 'package:mixeditems/sections/icon_detail/version_section.dart';
+
+class WidgetPlaceHolder {
```
Comment (mkoechig) [inline lib/repository/widget_placeholder.dart]: Why placeholder? This seems to be a 'factory', as it's creating Widgets.

Comment (mkoechig) [inline lib/repository/widget_placeholder.dart]: This class is placed inside the repository package. But it is creating Widgets and therefore views. Repositories should only handle data and not handle nor create Views.

File: lib/repository/widget_placeholder.dart [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  List<Widget> getIconDetailWidgets(IconItem iconItem){
+    return [
+      HeadSection(iconItem),
+      UsageSection('Usage'),
+      SettingsSection('App settings'),
+      VersionSection(iconItem)
+    ];
+  }
+
+}
```
Comment (mkoechig) [inline lib/repository/widget_placeholder.dart]: This class also has not been services by the auto formatter

File: lib/screens/icon_detail_widget.dart [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text('App Info'),
+      ),
+      body: _buildListView(context),
+    );
+  }
+
+  Widget _buildListView(BuildContext context) {
+    WidgetPlaceHolder _widgetPlaceHolder = WidgetPlaceHolder();
```
Comment (mkoechig) [inline lib/screens/icon_detail_widget.dart]: This should be a property of the class, instead of being newly created everytime this function is called

**Review by mkoechig — CHANGES_REQUESTED** [review]

File: lib/repository/repo.dart (hunk 5 above)
Comment (Lee-Yoonjoo) [reply lib/repository/repo.dart]: I am not sure so I just did Auto format :)

File: lib/screens/icon_detail_widget.dart (hunk 8 above)
Comment (Lee-Yoonjoo) [reply lib/screens/icon_detail_widget.dart]: I see I will move them in the widget class.

File: lib/repository/widget_placeholder.dart (hunk 7 above)
Comment (Lee-Yoonjoo) [reply lib/repository/widget_placeholder.dart]: I will make sure

File: lib/repository/widget_placeholder.dart (hunk 6 above)
Comment (Lee-Yoonjoo) [reply lib/repository/widget_placeholder.dart]: I will move them in its widget class.

File: lib/screens/icon_detail_widget.dart (hunk 8 above)
Comment (mkoechig) [reply lib/screens/icon_detail_widget.dart]: That's not what I meant, but it's okay for now. It's not the scope of this branch.

**Review by mkoechig — APPROVED** [review]

**Pull request merged**